### PR TITLE
[KernelGen] Add Triton top_k_per_row_prefill for DeepSeek V4

### PR DESCRIPTION
**Pull request by yzw1128**
## Summary

Add a Triton implementation of `top_k_per_row_prefill` for DeepSeek V4 sparse attention prefill phase. This replaces vLLM's `persistent_topk` CUDA kernel.

## Approach

- **In-place masking kernel**: Triton kernel masks invalid ranges with `-inf`, with early exit for full-vocab rows (most common case in inference)
- **Adaptive selection**:
  - `num_rows=1`: `torch.argsort` (CUB radix sort, O(N), fastest for single row)
  - `num_rows>1`: `torch.topk` with `sorted=False` (heap-based, O(N log k), fastest for batches)
- **Fused postprocess kernel**: Single Triton kernel for slice + cast + subtract (index adjustment relative to row_starts)

## Performance

Tested on DeepSeek V4 config (vocab_size=129280, top_k=1024):

| num_rows | Speedup vs vLLM CUDA |
|----------|---------------------|
| 1        | 0.89x               |
| 32       | 0.38x               |

The kernel is competitive for single-row (decode-like) scenarios. Multi-row performance is bounded by `torch.topk` overhead on large vocab.

## Files Changed

- `src/flag_gems/fused/top_k_per_row_prefill.py` — Triton kernel implementation
- `src/flag_gems/fused/__init__.py` — Export new operator
- `conf/operators.yaml` — Register operator
- `tests/test_top_k_per_row_prefill.py` — Accuracy tests
- `benchmark/test_top_k_per_row_prefill.py` — Performance benchmark

## Test Plan

- Correctness verified against `torch.topk` reference across:
  - Full vocab (row_starts=0, row_ends=vocab_size)
  - Variable row lengths
  - Non-zero row_starts
  - num_rows in [1, 32, 64, 2048]
- Value-based set comparison (order-independent)

### CODE REVIEW

File: src/flag_gems/fused/top_k_per_row_prefill.py [hunk 1]
```diff
@@ -0,0 +1,181 @@
+"""Triton top_k_per_row_prefill for DeepSeek V4 sparse attention.
```
Comment (Vincent-Xiao) [inline src/flag_gems/fused/top_k_per_row_prefill.py:1]: 移到 fused 目录

File: tests/test_top_k_per_row_prefill.py [hunk 2]
```diff
@@ -0,0 +1,190 @@
+"""Accuracy tests for top_k_per_row_prefill (DeepSeek V4 sparse attention).
```
Comment (Vincent-Xiao) [inline tests/test_top_k_per_row_prefill.py:1]: 移到tests 目录下

**Review by Vincent-Xiao — COMMENTED** [review]
目录规范

File: benchmark/test_top_k_per_row_prefill.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+The torch reference uses torch.topk directly (no masking), representing the
+theoretical minimum for selection-only. The Triton kernel additionally handles
+masking and index adjustment.
+"""
+
+import pytest
+import torch
+
+import flag_gems
+from flag_gems.fused.top_k_per_row_prefill import top_k_per_row_prefill
```
Comment (tengqm) [inline benchmark/test_top_k_per_row_prefill.py]: The symbol `top_k_per_row_prefill` is already exported from `flag_gems.fused`, you do not need to import it from the source file. And ... that is the reason we register `top_k_per_row_prefill` into `__all__` in `flag_gems/fused/__init__.py`.

Please change this to

```
from flag_gems.fused import top_k_per_row_prefill
```

File: benchmark/test_top_k_per_row_prefill.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+            # This under-counts the Triton kernel's work (which also masks + adjusts),
+            # but provides a consistent PyTorch-native comparison point.
+            _, top_idx = torch.topk(logits, top_k, dim=1, largest=True, sorted=False)
+            indices.copy_(top_idx.to(torch.int32))
+
+        return torch_topk_ref
+
+
+@pytest.mark.top_k_per_row_prefill
+def test_perf_top_k_per_row_prefill():
```
Comment (tengqm) [inline benchmark/test_top_k_per_row_prefill.py]: for any benchmark function under `benchmark/` directory, the test has to be related to performance, so please avoid the `__perf` thing in the function name.
It is useless and it renders the test report wordy and difficult to digest.

File: conf/operators.yaml [hunk 5]
```diff
@@ -889,6 +889,18 @@ ops:
       - NeuralNetwork
     stages:
       - beta: '5.1'
+  - id: top_k_per_row_prefill
```
Comment (tengqm) [inline conf/operators.yaml]: This file is sorted by operator IDs.
Please keep the order rather than insert something randomly in the middle.

File: src/flag_gems/fused/__init__.py [hunk 6]
```diff
@@ -101,6 +102,7 @@
     "sinkhorn_forward",
     "skip_layer_norm",
     "swiglu",
+    "top_k_per_row_prefill",
```
Comment (tengqm) [inline src/flag_gems/fused/__init__.py:109]: I'm wondering if the operator should be named as `topk_per_row_prefill`.

File: src/flag_gems/fused/top_k_per_row_prefill.py [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+
+Replaces vLLM's persistent_topk CUDA kernel with a Triton implementation.
```
Comment (tengqm) [inline src/flag_gems/fused/top_k_per_row_prefill.py:3]: Not sure if we need to add an operation in the patch_vllm tool for this new operator.

File: tests/test_top_k_per_row_prefill.py [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+        logits, row_starts, row_ends, indices_test, num_rows, stride0, stride1, top_k
+    )
+
+    assert check_topk_values_match(
+        logits, indices_test, indices_ref, row_starts, top_k
+    ), f"FAIL: num_rows={num_rows}, nonzero starts"
+
+
+if __name__ == "__main__":
+    pytest.main([__file__, "-v"])
```
Comment (tengqm) [inline tests/test_top_k_per_row_prefill.py]: Please avoid inventing a new benchmark suite.
We use `pytest`.

File: tests/test_top_k_per_row_prefill.py [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+)
+def test_top_k_per_row_prefill_variable_lengths(num_rows, vocab_size, top_k):
+    """Test with variable row lengths (partial vocab per row).
+
+    Simulates the case where different tokens in a batch have different valid
+    KV ranges (e.g., due to causal masking or sequence packing).
+    row_ends is randomized in [top_k, vocab_size] to ensure enough valid elements.
+    """
+    if top_k > vocab_size:
+        pytest.skip("top_k > vocab_size")
```
Comment (tengqm) [inline tests/test_top_k_per_row_prefill.py]: for illegal input combinations, use a simple `return` rather than invoking `skip`. `skip` is used for unsatisfied environments, hardware constraints or operator/test that deem to fail.

Comment (tengqm) [reply tests/test_top_k_per_row_prefill.py]: in addition to the above comment, this test case won't get trapped into the `top_k > vocab_size` situation.
A guardrail can be placed into the operator implementation rather than in the test case.

**Comment by bin913** [comment]
@yzw1128 please fix the failing checks

**Review by bin913 — APPROVED** [review]
